### ISSUE TITLE
Support for passing options to migration functions

### ISSUE THREAD

**Issue by ksmithut** [zookzook/elixir-mongodb-driver#164]
I'm currently working with a multi-tenant application where each tenant has their own database. We currently seed these databases on the fly when new tenants get created. One thing that's missing for us is the ability to run these migrations for a new tenant (and for all tenants) using the new migrations feature because the underlying calls to lock the migration collection/document aren't taking in options.

We'd be looking for something like this:

```elixir
Mongo.Migration.migrate(database: "tenant1_db")
```

and the migration file:

```elixir
defmodule Mongo.Migrations.AddIndexes do
  def up(opts) do
    indexes = [
      [key: [email: 1], name: "email_index", unique: true]
    ]

    Mongo.create_indexes(:my_db, "my_collection", indexes, opts)
  end

  def down(opts) do
    Mongo.drop_index(:my_db, "my_collection", "email_index", opts)
  end
end
```

This is probably a super naive approach because part of the expectation would be that the mongo calls to the migration collection would also respect those options, and I don't think every mongo function takes in the same options.

I'm happy to give it a go at a PR for something I had working in my project, though I really only tried it for my use case. Is this something you've already thought about? Or would you have an alternate solution to what I'm trying to do?

Main things that are important to me:
- The migration collection that keeps track of the current migration is tenant specific. That way if a migration fails at a tenant, it's not in this weird state where some tenants have migrated, some haven't, but it didn't roll back the tenants that already ran.
- It's easy to run a single tenant's migrations. I can take care of the looping through tenants and running the migration function, I just need a way to tell the migrations to run against a specific database.

If that's not something you're looking to support, I've got a workaround sort of working.

**Comment by zookzook** [comment]
Thank you for the answer. I think there is a need to support multiple topologies, databases, and options to make it more flexible. I am working on that right now. It will support additional options so you have full control over the migration.

**Comment by ksmithut** [comment]
That’s awesome :) let me know if you’d like help with anything. Happy to jump in

**Comment by zookzook** [comment]
Hey! Sorry for the delay. You can try out the main branch. I added support for options and multiple topologies.

**Comment by ksmithut** [comment]
That's pretty awesome :) I think that will be super helpful for a lot of use cases. I'm not sure it will work for our use case as we have a database per tenant with new tenants being created fairly regularly. Because the topology names are atoms, that's an ever-increasing number of atoms. I'd be worried about the process footprint as well spinning up those additional gen_servers, especially for something as simple (in our use case) as a database parameter being passed through to various Repo operations. It's possible our use case just might be too complex, or we might need to create our own solution.

So far, here is what we've had to do to get our multitenant setup working:
- We put our tenant information (database name, etc) on the process dictionary.
- We copied over all of the `Mongo.Repo` functions over to our own `Mongo.Repo` module
- We had to modify a few of those functions to accepts mongo options, and pull off the database from the process dictionary to input it into the lower-level mongo calls.
- For migrations, we also copied it over and modified the functions to take in options such as the database name, which could also be forwarded onto the generated migration files themselves.

One thought I've had was to submit a PR for the `Repo` module functions to make sure they all accepted options and that those options could be passed through to the underlying mongo function. Then provide some kind of hook someone could use to modify options before they get passed into the mongo function. In our use case, it would inject the `:database` option from the process dictionary. If that's something that interests you, I can work up a PR for you to consider, bounce ideas off of. If that's not the direction you'd like to take this library, I totally understand :)

**Comment by zookzook** [comment]
The `Repo` module is in a very early stage, so PRs are welcome. The process dictionary "trick" is used for the transaction API as well. And I think the database option seems to be a good candidate to nest Repo calls. Thank you for your efforts!

**Comment by zookzook** [comment]
Should be fixed in version 1.0.1! Feel free to create a PR for the `:database` option if you are ready.
